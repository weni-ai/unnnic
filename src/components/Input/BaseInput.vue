<template>
  <input
    v-if="mask"
    v-mask="mask"
    v-bind="attributes"
    :value="fullySanitize(modelValue)"
    :class="[classes, { 'focus': forceActiveStatus }]"
    :type="nativeType"
    :readonly="readonly"
  />
  <input
    v-else
    v-bind="attributes"
    :value="fullySanitize(modelValue)"
    :class="[classes, { 'focus': forceActiveStatus }]"
    :type="nativeType"
    :maxlength="maxlength"
    :readonly="readonly"
  />
</template>

<script>
import { mask } from 'vue-the-mask';
import { fullySanitize } from '../../utils/sanitize';

export default {
  directives: { mask },
  props: {
    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },
    modelValue: {
      type: String,
      default: '',
    },
    nativeType: {
      type: String,
      default: 'text',
    },
    size: {
      type: String,
      default: 'md',
    },
    mask: {
      type: [String, Array],
      default: '',
    },
    hasIconLeft: Boolean,
    hasIconRight: Boolean,
    hasClearIcon: Boolean,
    maxlength: {
      type: Number,
      default: null,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    forceActiveStatus: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['update:modelValue'],
  data() {
    return {};
  },
  computed: {
    attributes() {
      return {
        ...this.$attrs,
        ...this.$attrs['v-bind'],
        onInput: (event) => {
          this.$emit('update:modelValue', event.srcElement.value);
        },
      };
    },

    classes() {
      return [
        'input',
        `size-${this.size}`,
        this.type,
        {
          'input--has-icon-left': this.hasIconLeft,
          'input--has-icon-right': this.hasIconRight,
          'input--has-clear-icon': this.hasClearIcon,
        },
      ];
    },
  },
  methods: {
    fullySanitize,
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;
@use './Input.scss' as *;

.input {
  @include input-base;

  box-sizing: border-box;
  width: 100%;

  &.size {
    &-md {
      @include input-md-font;
      padding: $unnnic-space-3 $unnnic-space-4;
    }

    &-sm {
      @include input-sm-font;
      padding: $unnnic-space-2 $unnnic-space-4;
    }
  }

  &.input--has-icon-left {
    padding-left: $unnnic-space-10;
  }

  &.input--has-icon-right {
    padding-right: $unnnic-space-10;
    &.input--has-clear-icon {
      padding-right: $unnnic-space-10 + $unnnic-space-6;
    }
  }

  &.input--has-clear-icon {
      padding-right: $unnnic-space-10;
    }

  &.error {
    @include input-error;
  }

  &:-ms-input-placeholder,
  &::-ms-input-placeholder {
    /* Internet Explorer 10-11 */
    color: $unnnic-color-fg-muted;
  }
}
</style>
