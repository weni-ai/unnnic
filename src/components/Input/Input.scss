@use '@/assets/scss/unnnic' as *;

@mixin input-base {
  background: $unnnic-color-white;
  border: $unnnic-border-width-thinner solid $unnnic-color-border-base;
  outline: none;
  border-radius: $unnnic-radius-2;
  color: $unnnic-color-fg-emphasized;
  caret-color: $unnnic-color-fg-muted;
  font: $unnnic-font-body;

  &:focus,
  &.focus {
    border-color: $unnnic-color-border-active;
  }

  &::placeholder {
    color: $unnnic-color-fg-muted;
    opacity: 1;
    /* Firefox */
  }

  &:disabled {
    border-color: $unnnic-color-border-muted;
    background-color: $unnnic-color-bg-muted;
    cursor: not-allowed;
    color: $unnnic-color-fg-muted;
  }
}

@mixin input-sm-font {
  font: $unnnic-font-body;
}

@mixin input-md-font {
  font: $unnnic-font-body;
}

@mixin input-error {
  border-color: $unnnic-color-border-critical;
}