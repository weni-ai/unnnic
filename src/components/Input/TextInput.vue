<template>
  <div
    :class="[
      'text-input',
      `size--${size}`,
      `text-input--icon-right-size-${iconRightSize}`,
    ]"
  >
    <BaseInput
      v-bind="attributes"
      ref="base-input"
      :modelValue="modelValue"
      :size="size"
      :nativeType="
        nativeType === 'password' && showPassword ? 'text' : nativeType
      "
      :type="type"
      class="input-itself"
      :hasIconLeft="!!iconLeft"
      :hasIconRight="!!iconRight || allowTogglePassword"
      @focus="onFocus"
      @blur="onBlur"
    />

    <UnnnicIcon
      v-if="iconLeft"
      :scheme="iconScheme"
      :icon="iconLeft"
      size="sm"
      :clickable="iconLeftClickable"
      :class="['icon-left', { clickable: iconLeftClickable }]"
      @click="onIconLeftClick"
    />

    <UnnnicIcon
      v-if="iconRightSvg"
      :scheme="iconScheme"
      :icon="iconRightSvg"
      :size="iconRightSize"
      :clickable="iconRightClickable || allowTogglePassword"
      :class="[
        'icon-right',
        { clickable: iconRightClickable || allowTogglePassword },
      ]"
      @click="onIconRightClick"
    />
  </div>
</template>

<script>
import BaseInput from './BaseInput.vue';
import UnnnicIcon from '../Icon.vue';

export default {
  components: {
    BaseInput,
    UnnnicIcon,
  },
  props: {
    placeholder: {
      type: String,
      default: null,
    },
    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },
    modelValue: {
      type: String,
      default: '',
    },
    nativeType: {
      type: String,
      default: '',
    },
    iconLeft: {
      type: String,
      default: null,
    },
    iconRight: {
      type: String,
      default: null,
    },
    iconLeftClickable: {
      type: Boolean,
      default: null,
    },
    iconRightClickable: {
      type: Boolean,
      default: null,
    },
    iconRightSize: {
      type: String,
      default: 'sm',
    },
    allowTogglePassword: {
      type: Boolean,
      default: null,
    },
    hasCloudyColor: {
      type: Boolean,
      default: null,
    },
    size: {
      type: String,
      default: 'md',
    },
  },
  emits: ['icon-left-click', 'icon-right-click'],
  data() {
    return {
      isFocused: false,
      showPassword: false,
    };
  },
  computed: {
    isDisabled() {
      return this.$attrs.disabled;
    },

    iconRightSvg() {
      if (this.allowTogglePassword) {
        return this.showPassword ? 'view-off-1' : 'view-1-1';
      }

      return this.iconRight;
    },

    iconScheme() {
      if (this.type === 'error') {
        return 'aux-red-500';
      }

      if (this.isDisabled) {
        return 'neutral-cleanest';
      }

      if (this.modelValue || this.isFocused) {
        return 'neutral-dark';
      }

      if (this.hasCloudyColor) {
        return 'neutral-cloudy';
      }

      return 'neutral-cloudy';
    },

    attributes() {
      return { ...this.$attrs, ...this.$attrs['v-bind'], ...this.$props };
    },
  },

  methods: {
    focus() {
      this.$refs['base-input'].$el.focus();
    },

    onFocus() {
      this.isFocused = true;
    },

    onBlur() {
      this.isFocused = false;
    },

    onIconLeftClick() {
      if (this.iconLeftClickable) this.$emit('icon-left-click');
    },

    onIconRightClick() {
      if (this.attributes.disabled) return;
      if (this.allowTogglePassword) this.showPassword = !this.showPassword;
      else if (this.iconRightClickable) this.$emit('icon-right-click');
    },
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;

.icon {
  &-left,
  &-right {
    &:not(.clickable) {
      pointer-events: none;
    }
  }

  &-left {
    position: absolute;
    top: $unnnic-spacing-ant + 0.1875 * $unnnic-font-size;
    left: $unnnic-inline-sm - $unnnic-border-width-thinner;
  }

  &-right {
    position: absolute;
    top: $unnnic-spacing-ant + 0.1875 * $unnnic-font-size;
    right: $unnnic-inline-sm - $unnnic-border-width-thinner;
  }
}
.text-input {
  position: relative;

  &.size--sm {
    .icon-left,
    .icon-right {
      top: $unnnic-spacing-xs + 0.125 * $unnnic-font-size;
    }
  }

  &.size--sm.text-input--icon-right-size-ant .icon-right {
    top: 0.5625 * $unnnic-font-size;
  }

  &.size--md.text-input--icon-right-size-ant .icon-right {
    top: 0.8125 * $unnnic-font-size;
  }
}
</style>
