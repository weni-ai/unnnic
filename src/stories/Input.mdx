import { Meta, Source, Story } from '@storybook/blocks';

import * as InputStories from './Input.stories';

<Meta of={InputStories}/>

# Input

The `Input` component is designed to provide an input field for users to enter texts or passwords.

<Source
  language='html'
  dark
  code={`<UnnnicFormElement
  label="Label"
  message="Helper text" | error="Error text"
  :disabled="false | true"
>
  <UnnnicInput
    v-model="value"
    size="md"
    type="normal"
    nativeType="normal"
    placeholder="Placeholder"
    iconLeft="storefront"
    iconRight="storefront"
  />
</UnnnicFormElement>`}
/>


---

#### **Props Options:**

| Key         | Description                              | Values                   | Default     |
|-------------|------------------------------------------|--------------------------|-------------|
| v-model     | `String` of the input content            | `''`                     | `''`        |
| size        | `String` of the size                     | `'md'` \| `'sm'`         | `md`        |
| type        | `String` of the input type               | `'normal'` \| `'error'`  | `'normal'`  |
| nativeType  | `String` of the native input type        | `'text'` \| `'password'` | `'text'`    |
| placeholder | `String` of the native input placeholder | `String`                 | `''`        |
| iconLeft    | `String` of the left icon                | `String`                 | `undefined` |
| iconRight   | `String` of the right icon               | `String`                 | `false`     |

## Example
Some examples of uses of the `UnnnicInput`

### Normal

<Story of={InputStories.Normal} />


### Error

<Story of={InputStories.Error} />


### Disabled

<Story of={InputStories.Disabled} />


### Normal (small size)

<Story of={InputStories.NormalSm} />


### Error (small size)

<Story of={InputStories.ErrorSm} />


### Disabled (small size)

<Story of={InputStories.DisabledSm} />
