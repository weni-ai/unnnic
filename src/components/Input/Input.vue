<template>
  <div :class="['unnnic-form', size]">
    <p
      v-if="hasLabelSlot"
      class="unnnic-form__label"
    >
      <slot name="label" />
    </p>

    <p
      v-else-if="label"
      class="unnnic-form__label"
    >
      {{ sanitizedLabel }}
    </p>

    <TextInput
      v-bind="$attrs"
      v-model="val"
      class="unnnic-form-input"
      :placeholder="sanitizedPlaceholder"
      :iconLeft="iconLeft"
      :iconRight="iconRight"
      :type="type"
      :iconLeftClickable="iconLeftClickable"
      :iconRightClickable="iconRightClickable"
      :hasCloudyColor="hasCloudyColor"
      :size="size"
      :mask="mask"
      :nativeType="nativeType"
    />

    <p
      v-if="message"
      class="unnnic-form__message"
    >
      {{ sanitizedMessage }}
    </p>
  </div>
</template>

<script>
import { escapeHtml } from '../../utils/sanitize';
import TextInput from './TextInput.vue';

export default {
  name: 'UnnnicInput',
  components: { TextInput },
  props: {
    placeholder: {
      type: String,
      default: '',
    },
    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },
    modelValue: {
      type: String,
      default: '',
    },
    nativeType: {
      type: String,
      default: 'text',
    },
    message: {
      type: String,
      default: null,
    },
    label: {
      type: String,
      default: null,
    },
    iconLeft: {
      type: String,
      default: null,
    },
    iconRight: {
      type: String,
      default: null,
    },
    allowTogglePassword: {
      type: Boolean,
      default: false,
    },
    iconLeftClickable: {
      type: Boolean,
      default: null,
    },
    iconRightClickable: {
      type: Boolean,
      default: null,
    },
    hasCloudyColor: {
      type: Boolean,
      default: false,
    },
    size: {
      type: String,
      default: 'md',
    },
    mask: {
      type: [String, Array],
      default: '',
    },
  },
  emits: ['update:modelValue'],
  data() {
    return {
      val: this.modelValue,
    };
  },
  computed: {
    hasLabelSlot() {
      return !!this.$slots.label;
    },
    sanitizedLabel(){
      return escapeHtml(this.label)
    },
    sanitizedMessage(){
      return escapeHtml(this.message)
    },
    sanitizedPlaceholder(){
      return escapeHtml(this.placeholder)
    }
  },
  watch: {
    val() {
      const sanitizedVal = escapeHtml(this.val)
      this.$emit('update:modelValue', sanitizedVal);
    },
    modelValue() {
      this.val = this.modelValue;
    },
  },
  mounted() {
    this.val = this.modelValue;
  },
};
</script>

<style lang="scss" scoped>
@import '../../assets/scss/unnnic.scss';

.unnnic-form {
  font-family: $unnnic-font-family-secondary;
  position: relative;

  &__message {
    font-size: $unnnic-font-size-body-md;
    margin: $unnnic-spacing-stack-nano 0;
    color: $unnnic-color-feedback-red;
  }

  &__label {
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-medium;
    font-size: $unnnic-font-size-body-gt;
    color: $unnnic-color-neutral-cloudy;
    margin: $unnnic-spacing-stack-xs 0;
  }

  &.sm &__label {
    font-size: $unnnic-font-size-body-md;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-medium;
  }
}
</style>
