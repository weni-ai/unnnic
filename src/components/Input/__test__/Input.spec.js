import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, afterEach, test } from 'vitest';
import Input from '@/components/Input/Input.vue';
import TextInput from '@/components/Input/TextInput.vue';

describe('Input.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(Input, {
      props: {
        placeholder: 'Enter text',
        type: 'normal',
        modelValue: '',
        nativeType: 'text',
        label: 'Sample Label',
        message: 'Error message',
        iconLeft: 'search',
        iconRight: 'clear',
        iconLeftClickable: true,
        iconRightClickable: true,
        allowTogglePassword: false,
        hasCloudyColor: false,
        size: 'md',
        mask: '####-####',
      },
      global: {
        components: {
          TextInput,
        },
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('renders label correctly', async () => {
    const label = wrapper.find('.unnnic-form__label');
    expect(label.exists()).toBe(true);
    expect(label.text()).toBe('Sample Label');

    await wrapper.setProps({ label: null });
    expect(wrapper.find('.unnnic-form__label').exists()).toBe(false);
  });

  test('renders message correctly', async () => {
    const message = wrapper.find('.unnnic-form__message');
    expect(message.exists()).toBe(true);
    expect(message.text()).toBe('Error message');

    await wrapper.setProps({ message: null });
    expect(wrapper.find('.unnnic-form__message').exists()).toBe(false);
  });

  test('applies the correct size class', async () => {
    expect(wrapper.classes()).toContain('md');

    await wrapper.setProps({ size: 'sm' });
    expect(wrapper.classes()).toContain('sm');
  });

  test('renders TextInput component', () => {
    const textInput = wrapper.findComponent(TextInput);
    expect(textInput.exists()).toBe(true);
  });

  test('emits update:modelValue when val changes', async () => {
    const textInput = wrapper.findComponent(TextInput);
    await textInput.setValue('new value');
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    expect(wrapper.emitted('update:modelValue')[0]).toEqual(['new value']);
  });

  test('watch modelValue updates val', async () => {
    const wrapper = mount(Input, {
      props: {
        modelValue: 'initial value',
      },
      global: {
        components: {
          TextInput,
        },
      },
    });

    expect(wrapper.vm.val).toBe('initial value');

    await wrapper.setProps({ modelValue: 'new value' });

    expect(wrapper.vm.val).toBe('new value');
  });

  test('hasLabelSlot computed property works correctly', async () => {
    expect(wrapper.vm.hasLabelSlot).toBe(false);

    wrapper = mount(Input, {
      slots: {
        label: '<span>Custom Label</span>',
      },
    });

    expect(wrapper.vm.hasLabelSlot).toBe(true);
  });

  test('correctly mounts with initial modelValue', async () => {
    wrapper = mount(Input, {
      props: {
        modelValue: 'initial value',
      },
    });
    expect(wrapper.vm.val).toBe('initial value');
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
