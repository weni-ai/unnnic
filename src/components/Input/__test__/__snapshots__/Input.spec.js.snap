// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input.vue > matches the snapshot 1`] = `
"<div data-v-d890ad85="" class="unnnic-form md">
  <section data-v-7f222291="" data-v-d890ad85="" class="unnnic-label unnnic-form__label">
    <p data-v-7f222291="" class="unnnic-label__label">Sample Label</p>
    <!--v-if-->
  </section>
  <div data-v-a0d36167="" data-v-d890ad85="" class="text-input size--md unnnic-form-input" mask="####-####"><input data-v-86533b41="" data-v-a0d36167="" class="unnnic-form-input input-itself input size-md normal input--has-icon-left input--has-icon-right unnnic-form-input input-itself" placeholder="Enter text" iconleft="search" iconright="clear" iconleftclickable="true" iconrightclickable="true" hascloudycolor="false" type="text" value=""><span data-v-26446d8e="" data-v-a0d36167="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--fg-base unnnic-icon-size--ant unnnic-icon__size--ant unnnic--clickable icon-left clickable" data-testid="material-icon" translate="no">search</span><span data-v-26446d8e="" data-v-a0d36167="" class="unnnic-icon material-symbols-rounded unnnic-icon-scheme--fg-base unnnic-icon-size--ant unnnic-icon__size--ant unnnic--clickable icon-right clickable" data-testid="material-icon" translate="no">clear</span></div>
  <section data-v-d890ad85="" class="unnnic-form__hints-container">
    <section data-v-d890ad85="" class="unnnic-form__message-container">
      <p data-v-d890ad85="" class="unnnic-form__message">Error message</p>
      <!--v-if-->
    </section>
    <!--v-if-->
  </section>
</div>"
`;
